nixos/modules/services/cluster/k3s/default.nix

```diff
@@ -33,15 +33,38 @@ in
 
     serverAddr = mkOption {
       type = types.str;
-      description = lib.mdDoc "The k3s server to connect to. This option only makes sense for an agent.";
+      description = lib.mdDoc ''
+        The k3s server to connect to.
+
+        Servers and agents need to communicate each other.
+        Read [the networking docs][1] to know how to configure the firewall.
+
+        [1]: https://rancher.com/docs/k3s/latest/en/installation/installation-requirements/#networking
```

Comment:
individual descriptions are not independent markdown instances, using references like this could lead to problems. (this is the first one to appear at all)

Reply:
Maybe we could have an option to open firewall ports for K3s.

Reply:
I'd like to see an `openFirewall` option here. If it helps, I've just recently setup an HA k3s cluster and these are the ports which had to be configured:
```nix
{
  networking.firewall = {
    allowedUDPPorts = [
      53 # dns
      8472 # cni (flannel vxlan)
    ];
    allowedTCPPorts = [
      10250 # metrics server
    ] ++ (lib.optionals (cfg.role == "server") [
      6443 # k8s api
      2379 # etcd client requests
      2380 # etcd peer communication
    ]);
  };
}
```

Reply:
I deliberately didn't add that option because it depends a lot on how you want to configure the firewall.

The linked docs explain that some ports you opened shouldn't be world-reachable. Instead, most ports should be only reachable among nodes. ⚠️ Beware if that's the setup you're deploying in production!

To properly set up the firewall connections, we need to know other nodes' IPs. That fits better in multi-node deployment tools such as nixops or similar.

So, for now, I'll leave the firewall setup to the sysadmin.